### ISSUE TITLE
Add the missing routing primitives to the operation catalog

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#103]
> Filed against the `feature_request.yml` template. Category: **New Tensor Operation / Math Primitive**.

## Problem

Eight operations a mixture-of-experts layer needs are absent from the 174-operation catalog entirely, not merely missing on a backend. Verified against the generated operation list:

| Operation | Status | Why routing needs it |
|---|---|---|
| `one_hot` | **absent** | Turns expert indices into the dispatch matrix |
| `sort` | **absent** (only `argsort` exists) | Grouping tokens by expert needs the values, not only the permutation |
| `nonzero` | **absent** | Extracting the token indices assigned to one expert |
| `repeat_interleave` | **absent** | Replicating a token once per selected expert when `k > 1` |
| `bincount` | **absent** | Counting tokens per expert, for the load-balancing loss and for offsets |
| `scatter_add` | **absent** (`scatter` exists, **overwrite** semantics) | Combining expert outputs, where a token routed to `k` experts accumulates |
| `log_softmax` | **absent** | Numerically stable router log-probabilities |
| `logsumexp` | **absent** | The normalizer the auxiliary loss is built on |

Checked and **present**, so nobody re-derives this: `concat`, `where_cond`, `masked_fill`, `tril`, `triu`, `cumsum`, `topk`, `argsort`, `gather`, `scatter`, `index_select`, `softmax`.

`scatter_add` is the sharpest omission. `scatter` overwrites, so combining top-`k` expert outputs with it **silently keeps one contribution and discards the rest**. That is a wrong answer, not a missing feature, and it is the kind of wrong answer that looks like a convergence problem.

These are not MoE-only. `one_hot`, `log_softmax`, `nonzero`, and `bincount` are operations any user expects, and their absence is noticeable well before anyone builds a router.

## Proposed API sketch

Lines marked `// proposed` are new surface; the rest are existing catalog
operations, shown so the composition is legible end to end.

```rust
use incin::prelude::*;
type Backend = DefaultBackend;

const T: usize = 1024;   // tokens
const E: usize = 8;      // experts
const K: usize = 2;      // top-k

let logits: Tensor<s![T, E], Backend, f32, Grad> = router.forward(x)?;

// ── stable router probabilities ──
let logp = logits.log_softmax(1)?;                       // proposed: [T, E]
let norm  = logits.logsumexp(1, false)?;                 // proposed: [T]

// ── selection (topk exists; widened past f32 by #87) ──
let (weights, indices) = logits.topk(K, 1)?;             // [T, K] values, [T, K] i64

// ── dispatch ──
let dispatch = indices.one_hot::<E>()?;                  // proposed: [T, K, E] bool
let counts   = indices.bincount(E)?;                     // proposed: [E] i64
let offsets  = counts.cumsum(0)?;                        // exists: [E]
let order    = indices.flatten(..)?.argsort(0, false)?;  // exists: [T*K]
let expanded = x.repeat_interleave(K, 0)?;               // proposed: [T*K, D]
let grouped  = expanded.index_select(&order, 0)?;        // exists: [T*K, D]

// ── grouped matmul: one launch for all experts, ranges given by offsets ──
let hidden = grouped.grouped_matmul(&expert_weights, &offsets)?;   // proposed

// ── combine: MUST accumulate, not overwrite ──
let out = Tensor::<s![T, D_MODEL], Backend, f32, Grad>::zeros(())?
    .scatter_add(&inverse_order, &hidden_weighted, 0)?;            // proposed
```

## Shape and type-system invariants

| Operation | Shape rule | Notes |
|---|---|---|
| `one_hot::<N>` | `[..]` -> `[.., N]` | `N` is a const parameter so the output shape stays static. A runtime `N` would force `Dyn`. |
| `bincount(n)` | `[..]` -> `[n]` | Same reasoning: const `n` keeps the output static. |
| `sort(dim, desc)` | shape-preserving, returns `(values, indices)` | Stability guarantee must be documented. |
| `repeat_interleave(k, dim)` | `[.., D, ..]` -> `[.., D * k, ..]` | `k` const where possible; the extent arithmetic is ordinary shape algebra. |
| `scatter_add(index, src, dim)` | shape-preserving on `self` | Accumulates. Determinism is the contract question below. |
| `log_softmax(dim)` | shape-preserving | |
| `logsumexp(dim, keepdim)` | reduces `dim` | Follows the existing `*_dim` / `*_keepdim` convention. |
| **`nonzero`** | **data-dependent extent** | This is the hard one. See below. |
| `grouped_matmul(w, offsets)` | `[R, D] x [E, D, H]` with `[E+1]` offsets -> `[R, H]` | `R` is static; the per-expert split is a row range, not a shape. |

**`nonzero` is the exception and needs a decision.** Its output extent depends on the data, so it cannot have a static shape. It is the clearest existing instance of the question #102 has to answer, and it should be specified consistently with whatever that decides. The likely answer is `Tensor<Dyn, B, i64>` with the rank static and the extent dynamic.

## Background

- **The offsets trick is why this list is short.** In the sorted-buffer strategy (#102, option C), `n_e` never becomes a tensor extent. It becomes a row range inside a statically shaped `[T*K, D]` buffer, described by an `[E+1]` offsets array. `bincount` plus `cumsum` produce that array, and `grouped_matmul` consumes it. That is the whole mechanism, and it is why "MoE needs dynamic shapes" is only half true.
- **`scatter_add` determinism is a real contract, not boilerplate.** Floating-point accumulation over a repeated index set is order-dependent. A GPU implementation using atomics is fast and nondeterministic; a sorted-segment implementation is deterministic and slower. The choice must be stated in the row, and it must match between CPU and CUDA or the conformance harness (#83) will report a false failure.
- **`log_softmax` is not `softmax` then `log`.** Composing them loses precision exactly where router logits are extreme, which is where routing decisions are made. Same for `logsumexp`.
- **`sort` versus `argsort`.** `argsort` returns the permutation; `sort` returns the values too. Grouping needs both, and computing `sort` as `argsort` then `gather` costs an extra full materialization on the routing hot path.
- **Tie-breaking must be specified.** `topk`, `argsort`, and `sort` all have ties. CPU and CUDA agreeing by accident is not the same as agreeing, and #87 raises the same point for the existing operations.

## How this interoperates

- **`OPERATION_CATALOG` is the entry point.** Each operation needs a row in `crates/incin-core/src/exec/catalog/table.rs` with its full semantic contract: `BroadcastingRule`, `DTypeRule`, `OutputRule`, `EmptyRule`, `NumericRule`, `GradientRule`, `deterministic`, `LayoutRule`, and `ExecutionSite`. `execution_site` defaults to `ExecutionSite::Kernel`, which is the fail-closed answer, so a new row automatically owes an executor.
- **Capability declarations follow.** A CPU executor plus a row in `CPU_CAPABILITIES` (`crates/incin-backends/src/capability/tables.rs`). The CPU declaration generates a compile-time obligation that each advertised operation has an `Execute` impl, so an advertised-but-unimplemented row does not build.
- **Generated docs regenerate.** `docs/operation-coverage.md` (currently "Canonical operations: 174") and `docs/capabilities.md` are generated. Regenerate, do not edit.
- **#102** decides `nonzero`'s output contract and whether the grouped path is the target. Everything else here is independent of that decision.
- **#87 and #88** bring the *existing* routing primitives (`topk`, `cumsum`, `argsort`, `gather`, `scatter`, `index_select`) to CUDA. This issue plus those two is the complete routing surface.
- **#85** provides the batched GEMM that `grouped_matmul` extends with per-expert offsets.
- **#101** benefits independently: `log_softmax` is what a language-model head should use.

## Acceptance criteria

- All eight operations exist in `OPERATION_CATALOG` with complete semantic rows and CPU executors.
- `scatter_add` accumulates repeated indices correctly, with a documented and tested determinism guarantee that CPU and CUDA both honor.
- `nonzero`'s dynamic output extent is expressible through the tensor API rather than only through backend internals.
- Tie-breaking for `sort` is documented in `docs/OPERATION_SEMANTICS.md` and matches `argsort`/`topk`.
- Gradients are defined and tested for every operation that has one; operations without one say so in their catalog row.
- `grouped_matmul` executes on CPU and CUDA and matches a loop of individual matmuls, including the empty-expert case where a range is zero rows.
- `docs/operation-coverage.md` and `docs/capabilities.md` are regenerated, not edited.

## Alternatives

- **Compose from existing operations.** `one_hot` is expressible as `cmp_eq` against an `arange`. It costs an extra materialization on the routing hot path, and it does not work at all for `nonzero`, `bincount`, or `scatter_add`, which are the three that matter most.
- **Add only what MoE strictly needs**, skipping `sort` and `logsumexp`. Saves two rows and leaves the catalog with a shape users will keep tripping over.
- **`scatter` with an `accumulate: bool` attribute** rather than a separate `scatter_add` operation. Fewer rows; makes a semantics change look like a flag, and makes the determinism contract conditional on an attribute value.

## Additional context

Relevant code: `crates/incin-core/src/exec/catalog/table.rs` (`OperationCatalogEntry`, `execution_site`), `crates/incin-core/src/exec/catalog/classification.rs` (`DTypeRule`, `ExecutionSite`), `crates/incin-backends/src/cpu/ops/`, `crates/incin-backends/src/capability/tables.rs`, `docs/OPERATION_SEMANTICS.md`.

Independent of #102 except for `nonzero`'s output contract. CUDA implementations of the existing primitives are in #87 and #88. `grouped_matmul` depends on #85.

**Comment by xupremix** [comment]
one_hot landed on develop (commit cd5b9232): index.one_hot::<E>() with the depth as a const parameter, CPU executor, capability row, conformance fixture, and descriptor-first refusal of float operands. Out-of-range indices encode as all-false rows per ONNX. Still absent from the catalog: sort, nonzero, repeat_interleave, bincount, grouped_matmul. The research branch that held the 0.2.0 analysis sweep has been merged as a dated record and deleted; its index now notes this slice.
